tutorials/tutorial-01.qmd

```diff
@@ -6,33 +6,1403 @@ time: "~45 minutes"
+# Fix any invalid geometries
+mpa_sf_list <- lapply(mpa_sf_list, st_make_valid)
+```
+
+#### Infrastructure
+
+```{r}
+#| label: validate-infrastructure-geometries
+sapply(infra_sf_list, function(x) sum(!st_is_valid(x)))
+```
+
+```{r}
+#| label: fix-infrastructure-geometries
+infra_sf_list <- lapply(infra_sf_list, st_make_valid)
+```
+
+### Organising Data Structures
+
+Now we need to organise our data for analysis. The protected areas and infrastructure require different approaches:
+
+- **Protected areas**: We'll combine the three MPA layers into a single dataset. Our analysis question is "which protected areas intersect with infrastructure?" We want to test all MPAs together, while preserving a `type` column to distinguish their source. This requires standardising column names across layers.
+
+- **Infrastructure**: We'll keep pipelines, cables, and platforms as separate objects. Mixing different geometry types (points and lines) in a single `sf` object is awkward for both analysis and plotting, so it's more practical to keep them separate.
+
+#### Combining protected area layers
+
+The three MPA layers have different column names. For example, site names are stored as `sitename` in two layers but `name` in the third:
+
+```{r}
+#| label: inspect-protected-area-structures
+lapply(mpa_sf_list, names)
+```
+
+To combine them with `bind_rows()`, we need to standardise these first. We'll use the schema we defined in [Planning Our Data Schema](#planning-our-data-schema), where the `from` field maps source column names to our target names (e.g., `name` → `sitename`).
+
+Here's a helper function that applies these transformations:
+
+```{r}
+#| label: define-standardise-sf
+standardise_sf <- function(data, schema = NULL) {
+  # Type conversion map using readr-style shorthand
+  type_map <- list(
+    c = as.character,
+    i = as.integer,
+    d = as.double,
+    l = as.logical,
+    D = as.Date,
+    T = as.POSIXct
+  )
+
+  if (is.null(schema)) {
+    return(data)
+  }
+
+  # Build rename vector: target_name = source_name
+  # If 'from' not specified, use the target name (no rename)
+  rename_vec <- map_chr(names(schema), ~ schema[[.x]]$from %||% .x)
+  names(rename_vec) <- names(schema)
+
+  # Get type for each column, defaulting to "c" (character)
+  get_type <- function(col) type_map[[schema[[col]]$type %||% "c"]]
+
+  data |>
+    select(any_of(rename_vec)) |>
+    mutate(across(any_of(names(schema)), ~ get_type(cur_column())(.x)))
+}
+```
+
+Now we apply the schema to each layer, add a `type` label to track the source, and combine:
+
+```{r}
+#| label: standardise-and-combine-protected-areas
+# Standardise each layer and combine with type labels
+type_labels <- c(
+  cddaareas = "CDDA",
+  natura2000areas = "Natura 2000",
+  marineprotectedareas = "MPA"
+)
+
+mpa_combined <- imap(
+  mpa_sf_list,
+  ~ standardise_sf(.x, mpa_schemas[[.y]]$cols) |>
+    mutate(type = type_labels[[.y]])
+) |>
+  bind_rows()
+
+mpa_combined
```

Comment:
remove rownames? they are the first column in the table so the table doesn't look informative at first glance.